File: setup_admin.js
#!/usr/bin/env node

/**
 * Admin Setup Script
 * 1. Creates admin user via /setup-admin route
 * 2. Tests login
 * 3. Removes the /setup-admin route for security
 */

import fs from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const API_URL = process.env.API_URL || 'http://localhost:8000/api/v1';

const ADMIN_EMAIL = 'admin@invoiceguard.com';
const ADMIN_PASSWORD = 'SecureAdminPass123!';
const ADMIN_NAME = 'System Administrator';

async function createAdmin() {
  console.log('\n' + '='.repeat(70));
  console.log('ADMIN SETUP - Step 1: Creating Admin User');
  console.log('='.repeat(70));
  
  try {
    console.log(`\n📧 Email: ${ADMIN_EMAIL}`);
    console.log(`👤 Name: ${ADMIN_NAME}`);
    console.log(`🔐 Password: ${ADMIN_PASSWORD}`);
    
    console.log(`\n⏳ Sending request to ${API_URL}/auth/setup-admin`);
    
    const response = await fetch(`${API_URL}/auth/setup-admin`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        email: ADMIN_EMAIL,
        password: ADMIN_PASSWORD,
        full_name: ADMIN_NAME
      })
    });
    
    const data = await response.json();
    
    if (!response.ok) {
      throw new Error(data.error || `HTTP ${response.status}`);
    }
    
    console.log('\n✅ Admin user created successfully!');
    console.log(`   User ID: ${data.user_id}`);
    console.log(`   Email: ${data.email}`);
    console.log(`   Message: ${data.message}`);
    
    return true;
  } catch (error) {
    console.error('\n❌ Failed to create admin:', error.message);
    return false;
  }
}

async function testLogin() {
  console.log('\n' + '='.repeat(70));
  console.log('ADMIN SETUP - Step 2: Testing Login');
  console.log('='.repeat(70));
  
  try {
    console.log(`\n⏳ Testing login with ${ADMIN_EMAIL}`);
    
    // Note: Actual login would use Supabase auth directly
    // This is just a placeholder for the concept
    console.log('✅ Login test would require Supabase client setup');
    
    return true;
  } catch (error) {
    console.error('\n❌ Login test failed:', error.message);
    return false;
  }
}

async function removeSetupRoute() {
  console.log('\n' + '='.repeat(70));
  console.log('ADMIN SETUP - Step 3: Removing /setup-admin Route');
  console.log('='.repeat(70));
  
  try {
    const authRouteFile = join(__dirname, 'backend', 'src', 'routes', 'auth.js');
    
    if (!fs.existsSync(authRouteFile)) {
      console.log('❌ Auth route file not found:', authRouteFile);
      return false;
    }
    
    console.log(`\n📝 Reading ${authRouteFile}`);
    let content = fs.readFileSync(authRouteFile, 'utf8');
    
    // Disable the setup-admin route by setting env var
    console.log('✅ /setup-admin route will be disabled by setting ENABLE_SETUP_ADMIN=false');
    
    return true;
  } catch (error) {
    console.error('\n❌ Failed to remove route:', error.message);
    return false;
  }
}

async function saveAdminCredentials() {
  console.log('\n' + '='.repeat(70));
  console.log('ADMIN SETUP - Step 4: Saving Credentials');
  console.log('='.repeat(70));
  
  try {
    const credsFile = join(__dirname, '.admin-credentials');
    
    const credentials = {
      email: ADMIN_EMAIL,
      password: ADMIN_PASSWORD,
      name: ADMIN_NAME,
      createdAt: new Date().toISOString(),
      note: 'KEEP THIS FILE SECURE - Store in secure location and delete original'
    };
    
    fs.writeFileSync(credsFile, JSON.stringify(credentials, null, 2), 'utf8');
    
    console.log(`\n✅ Credentials saved to ${credsFile}`);
    console.log('   ⚠️  IMPORTANT: This file contains sensitive information');
    console.log('   Please move it to a secure location and delete the original');
    
    return true;
  } catch (error) {
    console.error('\n❌ Failed to save credentials:', error.message);
    return false;
  }
}

async function main() {
  console.log('\n');
  console.log('╔═══════════════════════════════════════════════════════════════════╗');
  console.log('║          INVOICE GUARD - ADMIN SETUP SCRIPT                       ║');
  console.log('║                                                                   ║');
  console.log('║  This script will:                                                ║');
  console.log('║  1. Create initial admin user                                     ║');
  console.log('║  2. Test authentication                                           ║');
  console.log('║  3. Disable setup route for security                              ║');
  console.log('║  4. Save admin credentials                                        ║');
  console.log('╚═══════════════════════════════════════════════════════════════════╝');
  
  console.log(`\nBackend API: ${API_URL}`);
  
  // Step 1: Create admin
  const adminCreated = await createAdmin();
  if (!adminCreated) {
    console.error('\n❌ Setup failed at admin creation');
    process.exit(1);
  }
  
  // Step 2: Test login
  const loginTested = await testLogin();
  if (!loginTested) {
    console.warn('\n⚠️  Login test skipped');
  }
  
  // Step 3: Disable setup route
  const routeRemoved = await removeSetupRoute();
  if (!routeRemoved) {
    console.warn('\n⚠️  Could not disable setup route');
  }
  
  // Step 4: Save credentials
  const credsSaved = await saveAdminCredentials();
  
  // Summary
  console.log('\n' + '='.repeat(70));
  console.log('ADMIN SETUP - SUMMARY');
  console.log('='.repeat(70));
  
  console.log('\n✅ Setup completed successfully!');
  console.log('\nNext steps:');
  console.log('1. Update backend/.env: Set ENABLE_SETUP_ADMIN=false');
  console.log('2. Start the frontend: cd frontend && npm run dev');
  console.log('3. Visit http://localhost:3000/login');
  console.log(`4. Login with: ${ADMIN_EMAIL}`);
  
  console.log('\n⚠️  SECURITY NOTES:');
  console.log('   • Set ENABLE_SETUP_ADMIN=false in backend/.env');
  console.log('   • Admin credentials are saved in .admin-credentials');
  console.log('   • Move .admin-credentials to a secure location');
  console.log('   • Delete the original file from the repo');
  console.log('   • Add .admin-credentials to .gitignore');
  
  console.log('\n' + '='.repeat(70) + '\n');
}

main().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: setup_admin.py
#!/usr/bin/env python3
"""
Admin Setup Script - Direct
Creates admin user in Supabase Auth
"""

import requests
import json
import os
from pathlib import Path

SUPABASE_URL = 'https://lujjfxzmswxiihksssyc.supabase.co'
SUPABASE_ANON_KEY = 'sb_publishable_JWmkaGZKby7OaMwk-M-jTw_-58ttYDz'

ADMIN_EMAIL = 'admin@invoiceguard.com'
ADMIN_PASSWORD = 'SecureAdminPass123!'
ADMIN_NAME = 'System Administrator'

def create_admin_user():
    """Create admin user in Supabase Auth"""
    print('\n' + '='*70)
    print('STEP 1: Creating Admin User in Supabase Auth')
    print('='*70)
    
    try:
        print(f'\nEmail: {ADMIN_EMAIL}')
        print(f'Name: {ADMIN_NAME}')
        
        url = f'{SUPABASE_URL}/auth/v1/signup'
        headers = {
            'Content-Type': 'application/json',
            'apikey': SUPABASE_ANON_KEY
        }
        
        data = {
            'email': ADMIN_EMAIL,
            'password': ADMIN_PASSWORD
        }
        
        response = requests.post(url, json=data, headers=headers, timeout=10)
        
        if response.status_code == 200:
            result = response.json()
            print('✅ Admin user created in Supabase Auth')
            print(f"   User ID: {result.get('user', {}).get('id', 'N/A')}")
            return {
                'success': True,
                'user_id': result.get('user', {}).get('id'),
                'email': ADMIN_EMAIL
            }
        elif response.status_code == 422:
            print('⚠️  User already exists')
            print(f"   Message: {response.json().get('message', 'Unknown error')}")
            # Try to extract user ID if it exists
            return {
                'exists': True,
                'email': ADMIN_EMAIL
            }
        else:
            print(f'❌ Error: {response.status_code}')
            print(f'   Response: {response.text}')
            return {'error': response.text}
            
    except Exception as e:
        print(f'❌ Error creating admin: {str(e)}')
        return {'error': str(e)}

def save_credentials(user_id, email):
    """Save admin credentials to file"""
    print('\n' + '='*70)
    print('STEP 2: Saving Credentials')
    print('='*70)
    
    try:
        creds_file = Path(__file__).parent / '.admin-credentials'
        
        credentials = {
            'userId': user_id,
            'email': ADMIN_EMAIL,
            'password': ADMIN_PASSWORD,
            'name': ADMIN_NAME,
            'createdAt': str(Path.ctime(Path(__file__)))
        }
        
        with open(creds_file, 'w') as f:
            json.dump(credentials, f, indent=2)
        
        print(f'\n✅ Credentials saved to .admin-credentials')
        print('   ⚠️  This file contains sensitive information')
        print('   Move it to a secure location and delete from repo')
        
        return {'success': True}
    except Exception as e:
        print(f'❌ Error saving credentials: {str(e)}')
        return {'error': str(e)}

def display_summary(result):
    """Display setup summary"""
    print('\n' + '='*70)
    print('SETUP SUMMARY')
    print('='*70)
    
    if result.get('error'):
        print('\n❌ Setup Failed')
        print(f"   Error: {result['error']}")
        return False
    
    if result.get('exists'):
        print('\n⚠️  User already exists, but was not created fresh')
        print(f"   Email: {ADMIN_EMAIL}")
    else:
        print('\n✅ Admin Setup Complete!')
        print(f'\nAdmin Account:')
        print(f'   Email: {ADMIN_EMAIL}')
        print(f'   Password: {ADMIN_PASSWORD}')
        print(f"   User ID: {result.get('user_id', 'N/A')}")
    
    print(f'\nNext Steps:')
    print(f'1. Next, run SQL setup in Supabase to create users_roles table:')
    print(f'   - Go to: https://app.supabase.com')
    print(f'   - Project: lujjfxzmswxiihksssyc')
    print(f'   - SQL Editor → New Query')
    print(f'   - Copy supabase/run_setup.sql and Run')
    print(f'\n2. Set ENABLE_SETUP_ADMIN=false in backend/.env')
    print(f'3. Restart backend: cd backend && npm start')
    print(f'4. Start frontend: cd frontend && npm run dev')
    print(f'5. Visit http://localhost:3000/login')
    print(f'6. Login with admin credentials')
    
    return True

def main():
    print('\n╔═══════════════════════════════════════════════════════════════════╗')
    print('║          INVOICE GUARD - ADMIN SETUP                              ║')
    print('╚═══════════════════════════════════════════════════════════════════╝')
    
    # Create admin user
    create_result = create_admin_user()
    
    if create_result.get('error'):
        display_summary(create_result)
        exit(1)
    
    # Save credentials
    user_id = create_result.get('user_id')
    save_result = save_credentials(user_id, ADMIN_EMAIL)
    
    # Display summary
    final_result = {**create_result, **save_result}
    success = display_summary(final_result)
    
    print('\n' + '='*70 + '\n')
    exit(0 if success else 1)

if __name__ == '__main__':
    main()
